march_hardware/src/PDOmap.cpp

```diff
@@ -6,129 +6,151 @@
+
+  for (const auto& nextObject : sorted_PDO_objects)
   {
-    // Check if register is still empty
-    if (sizeleft == this->bitsPerReg)
+    sizeLeft -= nextObject.second.length;
+    if (sizeLeft < 0)
     {
-      sdo_bit32(slaveIndex, reg, 0, 0);
+      // PDO is filled so it can be enabled again
+      sdo_bit8(slaveIndex, currentReg, 0, counter - 1);
+
+      // Update the sync manager with the just configured PDO
+      sdo_bit8(slaveIndex, baseSyncManager, 0, 0);
+      int currentPDONr = (currentReg - baseRegister) + 1;
+      sdo_bit16(slaveIndex, baseSyncManager, currentPDONr, currentReg);
+
+      // Move to the next PDO register by incrementing with one
+      currentReg++;
+      if (currentReg > (baseRegister + nr_of_regs))
+      {
+        ROS_ERROR("Amount of registers was overwritten, amount of parameters does not fit in the PDO messages.");
+      }
+
+      sizeLeft = this->bits_per_register - nextObject.second.length;
+      counter = 1;
+
+      sdo_bit8(slaveIndex, currentReg, 0, 0);
     }
-    // Get next object (from end, because sorted from small to large)
-    std::pair<IMCObjectName, IMCObject> nextObject = this->sortedPDOObjects.back();
-    this->sortedPDOObjects.pop_back();
-    // Add next object to map
+
+    int byteOffset = (bits_per_register - (sizeLeft + nextObject.second.length)) / 8;
+    byte_offsets[nextObject.first] = byteOffset;
+
+    sdo_bit32(slaveIndex, currentReg, counter,
+              PDOmap::combineAddressLength(nextObject.second.address, nextObject.second.length));
     counter++;
-    sdo_bit32(slaveIndex, reg, counter,
-              this->combineAddressLength(nextObject.second.address, nextObject.second.length));
-    this->byteOffsets[nextObject.first] = byteOffset;
-    byteOffset += nextObject.second.length / 8;
-    sizeleft -= nextObject.second.length;
-    // Check if this was the last object of the list
-    if (this->sortedPDOObjects.size() == 0)
-    {
-      sdo_bit32(slaveIndex, reg, 0, counter);
-      lastFilledReg = reg;
-      reg++;
-    }
-    // else, check if register is full
-    else if (sizeleft <= 0)
-    {
-      sdo_bit32(slaveIndex, reg, 0, counter);
-      reg++;
-      counter = 0;
-      sizeleft = this->bitsPerReg;
-    }
   }
-  // For the unused registers, set count to zero
-  for (int i = reg; i < startReg + this->nrofRegs; i++)
-  {
-    sdo_bit32(slaveIndex, i, 0, 0);
-  }
-  // For all filled registers, set data to Sync Manager object
-  int count = 0;
-  for (int i = startReg; i <= lastFilledReg; i++)
+
+  // Make sure the last PDO and sync manager are activated
+  sdo_bit8(slaveIndex, currentReg, 0, counter - 1);
+  sdo_bit8(slaveIndex, baseSyncManager, 0, 0);
+  int currentPDONr = (currentReg - baseRegister) + 1;
+  sdo_bit16(slaveIndex, baseSyncManager, currentPDONr, currentReg);
+
+  // Explicitly disable PDO registers which are not used
+  currentReg++;
+  if (currentReg <= (baseRegister + nr_of_regs))
```

Comment:
I am breaking my head a bit thinking of off-by-one errors here. I think here it will work out, but it could be better I think if the if condition were changed to a less than. Say you use all four registers (say 0x1600-0x1603). Then the currentReg will be incremented to 0x1604. The if condition is true (equality). However the for loop will quit immediately since unusedRegister (=0x1604) is not smaller than baseRegister+nr_of_regs (= also 0x1604). So that could probably be improved.

Reply:
See my comment in my later review.